version: '3'

services:

  redis:
    image: redis:latest
    restart: on-failure
    expose:
      - "6379"

  db:
    image: postgres:16
    volumes:
      - postgres_data:/var/lib/postgresql/data
    env_file:
      - ./.env
    expose:
      - "5432"
    healthcheck:
      test: [ "CMD-SHELL", "pg_isready -d $${POSTGRES_DB} -U $${POSTGRES_USER}"]
      interval: 30s
      timeout: 10s
      retries: 5

  web:
    build: .
    command: bash -c "
      python manage.py migrate
      && python manage.py runserver 0.0.0.0:8000
      "
    volumes:
      - static_volume:/code/static
    ports:
      - "8000:8000"
    depends_on:
      - db
    env_file:
      - ./.env

  nginx:
    build:
      context: ./nginx
    ports:
      - "80:80"
    volumes:
      - ./nginx/nginx.conf:/etc/nginx/nginx.conf
      - static_volume:/app/staticfiles
    depends_on:
      - web

  celery:
    build: .
    tty: true
    command: poetry run celery -A config beat -l INFO
    restart: on-failure
    depends_on:
      - redis
      - db
      - web
    volumes:
      - .:/app
    env_file:
      - .env

  celery-beat:
    build: .
    tty: true
    command: poetry run celery -A config beat -l INFO
    restart: on-failure
    depends_on:
      - redis
      - db
      - web
    volumes:
      - .:/app
    env_file:
      - .env

volumes:
  postgres_data:
  static_volume:
